tracing-attributes/src/lib.rs

```diff
@@ -292,15 +303,102 @@ pub fn instrument(
         )
         .into()
     } else {
-        gen_body(&input, args, instrumented_function_name, None).into()
+        gen_function(&input, args, instrumented_function_name, Vec::new(), None).into()
+    }
+}
+
+// perform the necessary work for recent versions of async-trait
+fn instrument_handle_recent_async_trait(
+    input_fn: &ItemFn,
+    args: InstrumentArgs,
+    async_expr: &ExprAsync,
+    instrumented_function_name: String,
+) -> Stmt {
+    let async_attrs = &async_expr.attrs;
+    let async_block = &async_expr.block.clone();
+
+    // extract statements up to the '__async_trait' marker if any
+    let mut block_finder = AsyncTraitBlockFinder { block: None };
+    syn::visit::visit_block(&mut block_finder, async_block);
+    if let Some(async_trait_block) = block_finder.block {
+        let pre_statements = async_trait_block.pre_statements;
+        let post_statements = async_trait_block.post_statements;
+
+        // collect all the variable rebindings to be fixed in user-supplied fields
+        let mut variable_rebindings = Vec::with_capacity(pre_statements.len());
+        for stmt in pre_statements.iter() {
+            if let Stmt::Local(Local {
+                pat: Pat::Ident(PatIdent { ident, .. }),
+                init: Some((_, source_path)),
+                ..
+            }) = stmt
+            {
+                if let Expr::Path(source_path) = source_path.as_ref() {
+                    variable_rebindings.push((
+                        ident.clone(),
+                        Ident::new(&path_to_string(&source_path.path), source_path.span()),
```

Comment:
why does the path need to be transformed into an ident here?

Reply:
Also part of code that was pruned